### PR TITLE
Use L1 loss for Smooth L1 loss with beta=0

### LINKED ISSUE #96813: Incorrect backward for smooth_l1_loss when beta = 0 on GPU
### 🐛 Describe the bug

**Bug description:**

Incorrect backward (i.e. NaN gradients) for smooth_l1_loss on GPU when `beta = 0` and when inputs and targets are nearly identical (see different `eps` values in code below). The gradient is correct when computed on CPU. I didn't have this issue when using PyTorch1.10 (and I believe it also worked correctly for PyTorch1.12, but I'm not 100% certain).

**Example to reproduce:**
```python
import torch
import torch.nn.functional as F

for eps in (0.000, 1e-8, 1e-7, 1e-6, 1e-5):
    input = torch.randn(300, 256, requires_grad=True, device='cuda')
    target = input.detach() + eps

    loss = F.smooth_l1_loss(input, target, beta=0.0, reduction='sum')
    loss.backward()

    if eps == 0:
        print(f"{eps:.3f}: {input.grad.mean().item()}")
    else:
        print(f"{eps}: {input.grad.mean().item()}")
```
**Output:**
```
0.000: nan
1e-08: nan
1e-07: nan
1e-06: -1.0
1e-05: -1.0
```
**Expected output:**
```
0.000: -1.0
1e-08: -1.0
1e-07: -1.0
1e-06: -1.0
1e-05: -1.0
```

### Versions

PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Fedora Linux 37 (Thirty Seven) (x86_64)
GCC version: (cpicron-ruchba) 9.3.0
Clang version: 15.0.7 (Fedora 15.0.7-1.fc37)
CMake version: version 3.26.0-rc5

Python version: 3.10 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1050 Ti
Nvidia driver version: 525.89.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==1.13.1
[pip3] torch-cluster==1.6.0
[pip3] torch-geometric==2.2.0
[pip3] torch-scatter==2.1.0
[pip3] torch-sparse==0.6.16
[pip3] torchvision==0.14.1
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0           py310h7f8727e_0  
[conda] mkl_fft                   1.3.1           py310hd6ae3a3_0  
[conda] mkl_random                1.2.2           py310h00e6091_0  
[conda] numpy                     1.23.5          py310hd5efca6_0  
[conda] numpy-base                1.23.5          py310h8e6c178_0  
[conda] pyg                       2.2.0           py310_torch_1.13.0_cu117    pyg
[conda] pytorch                   1.13.1          py3.10_cuda11.7_cudnn8.5.0_0    pytorch
[conda] pytorch-cluster           1.6.0           py310_torch_1.13.0_cu117    pyg
[conda] pytorch-cuda              11.7                 h67b0de4_1    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] pytorch-scatter           2.1.0           py310_torch_1.13.0_cu117    pyg
[conda] pytorch-sparse            0.6.16          py310_torch_1.13.0_cu117    pyg
[conda] torchvision               0.14.1              py310_cu117    pytorch

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mruberry @jbschlosser @walterddr @mikaylagawarecki @ngimel

### PR DESCRIPTION
**Pull request by CedricPicron**
Fixes #96813.

Comments:

1. Wasn't able to test since tools/nightly.py does not allow for GPU build (and I don't want to build from scratch).
2. In theory, the bug (i.e. NaNs) can still occur when beta is very small (e.g. `beta=1e-50`), but not sure whether anybody cares.
3. Some checks within the smooth_l1_loss C++ code could be changed to check for `beta > 0` instead of `beta >= 0`.

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -5397,6 +5397,26 @@ def test_smoothl1loss_negative_beta_not_supported(self):
...
+        input = torch.randn(300, 256, requires_grad=True, device='cpu')
+        target = input.detach()
+
+        loss = F.smooth_l1_loss(input, target, beta=0.0, reduction='sum')
+        loss.backward()
+
+        grad_max_abs = input.grad.abs().max().item()
+        self.assertLessEqual(grad_max_abs, 1.0)
+
+    def test_smoothl1loss_backward_zero_beta_cuda(self):
```
[the remark this answers is not in the document]
Comment (CedricPicron) [reply test/test_nn.py:5410]: Should I also add @onlyCPU for the cpu test or not?

[the remark this answers is not in the document]
Comment (albanD) [reply test/test_nn.py:5411]: Not that is not necessary.

[the remark this answers is not in the document]
Comment (albanD) [reply test/test_nn.py:5411]: Ho this decorate is for the per-device tests.
Can you move your test to the per-Device class? It is the ones just above that take `(self, device)` as argument instead of just self. 
From there, you can have one test that use the device passed in instead of having one test for cpu and one test for cuda.

**Comment by albanD** [comment]
Let me know if you want me to send a commit moving the test around to the right place if you don't know how to do it.

**Comment by CedricPicron** [comment]
I've moved the test. Feel free to move it if you had another location in mind.

**Review by jbschlosser — APPROVED** [review]
LGTM, thanks for the update!
